File: connect4.java
import java.util.Scanner;
public class connect4 {
    
    public static void main(String[] args){
        char board[][] = new char[8][7];
        int num;
        char a;
        int turn = 1;
        int column = 0;
        int countx = 0;
        int counto = 0;
        int thingy;
        try(Scanner scan = new Scanner(System.in)){
            for(int i = 1;i<8;i++){
                num = i;
                a = (char)(num + '0');
                board[0][i-1] = a;
            }
            for(int i = 1;i<8;i++){
                for(int j = 0;j<7;j++){
                    board[i][j] = '-';
                }
            }
            for(int i = 0;i<15;i++){
                printBoard(board);
                turn(board,turn,column,scan);
                if(turn == 1){
                    turn++;
                }
                else{
                    turn--;
                }
                thingy = checkWin(board,countx,counto);
                if(thingy == 1){
                    System.out.println("Player 1 wins!");
                }
                else if(thingy == 2){
                    System.out.println("Player 2 wins!");
                }
            }
        }
    }

    public static void printBoard(char board[][]){
        for(int i = 0;i<8;i++){
            for(int j = 0;j<7;j++){
                System.out.print(board[i][j] + " ");
            }
            System.out.println("");
        }
    }
    public static void turn(char board[][],int turn,int column,Scanner scan){
        int thingy = 0;
        int counter = 0;
        while(thingy == 0){
            counter = 0;
            System.out.println("Player " + turn + " pick a column");
            column = scan.nextInt();
            for(int i = 7;i>0;i--){
                if(board[i][column-1] == '-'){
                    if(turn == 1){
                        board[i][column-1] = 'X';
                    }
                    else{
                        board[i][column-1] = 'O';
                    }
                    i = 1;
                    thingy = 1;
                }
                else{
                    counter++;
                }
            }
            if(counter == 7){
                System.out.println("column is full");
            }
        }
            
        
        
    }
    public static int checkWin(char board[][],int countx,int counto){
        countx = 0;
        counto = 0;
        // check vertical
        for(int i = 0;i<7;i++){
            for(int j = 1;j<5;j++){
                for(int q = 0;q<4;q++){
                    if(board[j+q][i] == 'X'){
                        countx++;
                    }
                    else if(board[j+q][i] == 'O'){
                        counto++;
                    }
                }
                if(countx == 4){
                    return 1;
                }
                else if(counto == 4){
                    return 2;
                }
                countx = 0;
                counto = 0;
            }
        }
        // horizontal
        // diagnal
        return 0;
    }
    
    

}

File: connect4.py
def print_board(board):
    for i in range(6):
        for j in range(7):
            print(board[i][j],end=' ')
        print()

def check_win(board):
    # columns
    for i in range(7):
        for j in range(3):
            if board[j][i] == board[j+1][i] and board[j+1][i] == board[j+2][i] and board[j+2][i] == board[j+3][i] and board[j][i] != "-":
                return True
    # rows
    for i in range(6):
        for j in range(4):
            if board[i][j] == board[i][j+1] and board[i][j+1] == board[i][j+2] and board[i][j+2] == board[i][j+3] and board[i][j] != "-":
                return True
    # diagnals
    for i in range(3):
        for j in range(6):
            if(j<=3):
                if(board[i][j] == board[i+1][j+1] and board[i+1][j+1] == board[i+2][j+2] and board[i+2][j+2] == board[i+3][j+3] and board[i][j] != "-"):
                    return True
            if(j>=3):
                if(board[i][j] == board[i+1][j-1] and board[i+1][j-1] == board[i+2][j-2] and board[i+2][j-2] == board[i+3][j-3] and board[i][j] != "-"):
                    return True

def turn(board, player_turn):
    print_board(board)
    player_choice = int(input("What column would you like to go in? (1-7): ")) - 1
    coin_placed = False
    for i in range(5,-1,-1):
        if board[i][player_choice] == '-' and player_turn == 1 and coin_placed == False:
            board[i][player_choice] = 'X'
            coin_placed = True
        elif board[i][player_choice] == '-' and player_turn == 2 and coin_placed == False:
            board[i][player_choice] = 'O'
            coin_placed = True
    if not coin_placed:
        print("Column is full")
        turn(board,player_turn)
            

def main():
    board = [['-'for i in range(7)]for j in range(6)]
    player_one_wins = 0
    player_two_wins = 0
    would_like_to_play = input("Would you like to play connect 4? (y/n): ").lower()
    has_won = False
    while would_like_to_play == 'y':
        player_turn = int(input("Who will be going first? (1 or 2): "))
        has_won = False
        while not has_won:
            turn(board, player_turn)
            if check_win(board):
                print_board(board)
                if player_turn == 1:
                    print("Player 1 wins!")
                    player_one_wins += 1
                else:
                    print("Player 2 wins!")
                    player_two_wins += 1
                print("Player one wins:", player_one_wins)
                print("Player two wins:", player_two_wins)
                would_like_to_play = input("Would you like to play again? (y/n): ").lower()
                has_won = True
                board = [['-' for i in range(7)]for j in range(6)]
            else:
                if player_turn == 1:
                    player_turn += 1
                else:
                    player_turn -= 1
    



if __name__ == "__main__":
    main()